src/lib.rs

```diff
@@ -282,44 +391,163 @@ where
+
+                        let batches_and_wal_ids = minor_flush(
+                            &mut *guard,
+                            base_fs,
+                            immutable_chunk_num,
+                            immutable_chunk_max_num,
+                            true,
+                        )
+                        .await;
+
+                        let mut res = match batches_and_wal_ids {
+                            Ok(Some((mut batches, recover_wal_ids))) => {
+                                // Mark compaction window before releasing lock
+                                guard.compaction_in_progress.store(true, Ordering::Release);
+                                // Release lock before heavy work
+                                drop(guard);
+                                let rollback_wal_ids = recover_wal_ids.clone();
+                                let compaction_result = compactor
+                                    .check_then_compaction(
+                                        Some(&batches[..]),
+                                        recover_wal_ids,
+                                        true,
+                                    )
+                                    .await;
+                                let mut g = mem_storage_task.write().await;
+                                if compaction_result.is_err() {
+                                    for item in batches.drain(..).rev() {
+                                        g.immutables.insert(0, item);
+                                    }
+                                    if let Some(ids) = rollback_wal_ids {
+                                        if let Some(existing) = &mut g.recover_wal_ids {
+                                            existing.extend(ids);
+                                        } else {
+                                            g.recover_wal_ids = Some(ids);
+                                        }
+                                    }
+                                }
+                                g.compaction_in_progress.store(false, Ordering::Release);
+                                drop(g);
+                                compaction_result
+                            }
+                            Ok(None) => compactor.check_then_compaction(None, None, true).await,
+                            Err(e) => {
+                                error!("[Minor Flush Error]: {}", e);
+                                Ok(())
+                            }
+                        };
+
                         if let Some(tx) = option_tx {
-                            if result.is_ok() {
-                                result = tx.send(()).map_err(|_| CompactionError::ChannelClose);
+                            if res.is_ok() {
+                                res = tx.send(()).map_err(|_| CompactionError::ChannelClose);
                             }
                         }
```

Comment:
```suggestion
                        if let Some(tx) = option_tx {
                            // Always notify the caller to avoid hanging flush() even on error
                            let _ = tx.send(());
                        }
```